File: src/ios/DeviceOrientation.m
#import "DeviceOrientation.h"
#import <Cordova/CDV.h>

@implementation DeviceOrientation
@synthesize callbackId;



-(void)deviceOrientationDidChange
{
    [self.commandDelegate evalJs:@"cordova.fireDocumentEvent('orientationupdate', null, true);" scheduledOnRunLoop:YES];
}

- (void)getOrientation:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;

    [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
    [[NSNotificationCenter defaultCenter] addObserver:self 
                                             selector:@selector(deviceOrientationDidChange) 
                                                 name:UIDeviceOrientationDidChangeNotification
                                               object:nil];

    NSMutableDictionary* returnInfo = [NSMutableDictionary dictionaryWithCapacity:4];
    NSString* orientationStr;
    UIDeviceOrientation orientation = [[UIDevice currentDevice] orientation];
    switch (orientation) {
        case 1:
            orientationStr = @"Portrait";
            break;
        case 2:
            orientationStr = @"Upside Down";
            break;
        case 3:
            orientationStr = @"Landscape Right";
            break;
        case 4:
            orientationStr = @"Landscape Left";
            break;
        case 5:
            orientationStr = @"Camera Facing Down";
            break;
        case 6:
            orientationStr = @"Camera Facing Up";
            break;
        default:
            orientationStr = @"Unknown";
            break;
    }
    [returnInfo setObject:orientationStr forKey:@"orientation"];

    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:returnInfo];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end

File: src/ios/DeviceOrientation.swift
import CoreMotion

@objc(DeviceOrientation) class DeviceOrientation : CDVPlugin {
    var motionManager: CMMotionManager!;
    var timer: Timer!
    var accelerometerData: CMAccelerometerData!
    var lastOrientation: String = "";

    @objc func update() {
        let newOrientation: String = getOrientation();
        self.commandDelegate!.evalJs("cordova.fireDocumentEvent('orientationupdate', {orientation:'" + newOrientation + "'}, true);");
    }

    @objc(startAccelerometerUpdates:) func startAccelerometerUpdates(command: CDVInvokedUrlCommand) {
        motionManager = CMMotionManager();
        motionManager.startAccelerometerUpdates();
        timer = Timer.scheduledTimer(timeInterval: 0.2, target: self, selector:  #selector(update), userInfo: nil, repeats: true);

        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK
        )
        self.commandDelegate!.send(
            pluginResult, 
            callbackId: command.callbackId
        )
    }

    @objc(stopAccelerometerUpdates:) func stopAccelerometerUpdates(command: CDVInvokedUrlCommand) {
        motionManager.stopAccelerometerUpdates();
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK
        )
        self.commandDelegate!.send(
            pluginResult, 
            callbackId: command.callbackId
        )
    }

    func getOrientation() -> String {
        accelerometerData = motionManager.accelerometerData
        
        if accelerometerData != nil {
            let x = accelerometerData.acceleration.x;
            let y = accelerometerData.acceleration.y;
            if x >= 0.33 {
                lastOrientation = "landscape-secondary";
            } else if x <= -0.33 {
                lastOrientation = "landscape-primary";
            } else if y <= -0.33 {
                lastOrientation = "portrait-primary";
            } else if y >= 0.33 {
                lastOrientation = "portrait-secondary";
            }
            if lastOrientation.isEmpty {
                lastOrientation = "portrait-primary";
            }
            return lastOrientation;
        }
        return "portrait-primary"
    }
}
